src/lib/waku_relay/index.ts

```diff
@@ -92,86 +87,62 @@ export class WakuRelay extends GossipSub {
 
   /**
    * Send Waku message.
-   *
-   * @param {WakuMessage} message
-   * @returns {Promise<void>}
    */
-  public async send(message: WakuMessage): Promise<void> {
-    const msg = message.encode();
-    await this.publish(this.pubSubTopic, msg);
+  public async send(
+    encoder: Encoder,
+    payload: Uint8Array,
+    timestamp?: Date
+  ): Promise<PublishResult> {
+    const msg = await encoder.encode(payload, timestamp);
+    return this.publish(this.pubSubTopic, msg);
   }
 
   /**
-   * Register a decryption key to attempt decryption of received messages.
-   * This can either be a private key for asymmetric encryption or a symmetric
-   * key. `WakuRelay` will attempt to decrypt messages using both methods.
-   *
-   * Strings must be in hex format.
+   * Add decoders to decode incoming messages.
    */
-  addDecryptionKey(
-    key: Uint8Array | string,
-    options?: { method?: DecryptionMethod; contentTopics?: string[] }
-  ): void {
-    this.decryptionKeys.set(hexToBytes(key), options ?? {});
+  addDecoders(decoders: Decoder[]): void {
+    decoders.forEach((dec) => {
+      pushOrInitMapSet(this.decoders, dec.contentTopic, dec);
+    });
   }
 
   /**
-   * Delete a decryption key that was used to attempt decryption of received
-   * messages.
-   *
-   * Strings must be in hex format.
+   * Delete a decoder.
    */
-  deleteDecryptionKey(key: Uint8Array | string): void {
-    this.decryptionKeys.delete(hexToBytes(key));
+  deleteDecoders(decoders: Decoder[]): void {
+    decoders.forEach((dec) => {
+      const currDecs = this.decoders.get(dec.contentTopic);
+      if (currDecs) {
+        currDecs.delete(dec);
+      }
+    });
   }
 
   /**
-   * Register an observer of new messages received via waku relay
-   *
-   * @param callback called when a new message is received via waku relay
-   * @param contentTopics Content Topics for which the callback with be called,
-   * all of them if undefined, [] or ["",..] is passed.
-   * @returns {void}
+   * Add an observer to process incoming messages on a given content topics.
    */
   addObserver(
```

Comment:
I am uncertain what is the most practical API:

1. Call `addDecoder` and then `addObserver`
2. `addObserver` takes the decoders in parameters.

A similar API would then apply to Filter and Store.